### PR TITLE
LPS-189504: PUT operation Empty relationship node does not reset items related to the entry.

### PR DESCRIPTION
**Pull request by sergiojimcos**
**What is new?**
- Implement `fetchRelatedModel` for Custom Object Defintions in ObjectEntry1MModelsRelated and `System1MModelsRelated`.
- Update 1-M parsers checking valid data structure when it is going to set null data.
- Update `DefaultObjectEntryManager` and `ObjectRelationshipExtensionProvider` to disassociate related object entries in a Many to One relationship.
- Create `fetchManyToOneSystemObjectEntry` in DefaultObjectEntryManager.
- Create some test cases in different scenarios

**How to test this?**

1.Create University object
2. Create Student object
3. Create many-to-many relationship universitiesStudents with University as parent
4. Create University entry and Student entry and relate them. Save universityId for following steps.
5. Execute putUniversity:

```curl
curl -X 'PUT' \
  'http://localhost:8080/o/c/universities/${existingUniversityEntryId}' \
  -H 'accept: application/json' \
  -u '${username}:${password}'
  -d '{
  "name": "Oxford_updated",
  "universitiesStudents": [
  ]
}'
```
`universitiesStudents` relationship should be emptied. 

**How to deploy it?**
- Deploy `object-api` and `object-service`
- Deploy `object-rest-api` and `object-rest-impl`

**JIRA Related Ticket**
https://liferay.atlassian.net/browse/LPS-189504

### CODE REVIEW

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 1]
```diff
@@ -554,48 +554,77 @@ public ObjectEntry fetchManyToOneObjectEntry(
...
-		if (!Objects.equals(
+		FromStep fromStep = DSLQueryFactoryUtil.selectDistinct(
+			ObjectEntryTable.INSTANCE);
+		ObjectField objectField = _objectFieldPersistence.fetchByPrimaryKey(
+			objectRelationship.getObjectFieldId2());
+
+		if (Objects.equals(
 				objectRelationship.getType(),
-				ObjectRelationshipConstants.TYPE_ONE_TO_MANY) ||
+				ObjectRelationshipConstants.TYPE_ONE_TO_MANY) &&
```
Comment (ccorreagg) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:578]: I think we are missing the previous validation when the object relationship is not one to many

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/SystemObjectEntry1toMObjectRelationshipElementsParserImpl.java [hunk 2]
```diff
@@ -47,6 +48,10 @@ public List<Map<String, Object>> parse(
 		if (objectRelationship.getObjectDefinitionId1() ==
 				objectDefinition.getObjectDefinitionId()) {
 
+			if (MapUtil.isEmpty(parseOne(value))) {
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/SystemObjectEntry1toMObjectRelationshipElementsParserImpl.java:51]: You are invoking the parseOne method twice. Maybe better to keep the result in a variable

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/DefaultObjectEntryManagerImpl.java [hunk 3]
```diff
@@ -880,12 +892,21 @@ private Map<String, String> _addAction(
...
+						relatedObjectDefinition, objectRelationship,
+						objectDefinition)) {
+
 					disassociateRelatedModels(
 						objectDefinition, objectRelationship, primaryKey,
 						relatedObjectDefinition,
 						dtoConverterContext.getUserId());
 				}
+				else {
+					_disassociateRelatedModel(
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/DefaultObjectEntryManagerImpl.java:905]: This is very confusing. I don't know the difference between the public method and the private method as boths are called the same way

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/ObjectEntry1toMObjectRelationshipElementsParserImpl.java [hunk 4]
```diff
@@ -49,6 +51,13 @@ public List<ObjectEntry> parse(
 		if (objectRelationship.getObjectDefinitionId1() ==
 				objectDefinition.getObjectDefinitionId()) {
 
+			if ((value instanceof Map) && MapUtil.isEmpty((Map<?, ?>)value)) {
+				return Collections.emptyList();
+			}
+			else if (Objects.isNull(value)) {
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/ObjectEntry1toMObjectRelationshipElementsParserImpl.java:57]: Maybe it makes more sense to verify if value == null before the first if clause

File: modules/apps/object/object-rest-api/src/main/java/com/liferay/object/rest/manager/v1_0/DefaultObjectEntryManager.java [hunk 5]
```diff
@@ -76,6 +77,12 @@ public ObjectEntry fetchRelatedManyToOneObjectEntry(
 			String objectRelationshipName)
 		throws Exception;
 
+	public BaseModel<?> fetchRelatedManyToOneSystemObjectEntry(
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-api/src/main/java/com/liferay/object/rest/manager/v1_0/DefaultObjectEntryManager.java:80]: We are returning here a BaseModel, but in the rest of methods we return `Object`. Does it make sense?

Comment (ccorreagg) [reply modules/apps/object/object-rest-api/src/main/java/com/liferay/object/rest/manager/v1_0/DefaultObjectEntryManager.java:80]: This method is not being used from anywhere. Do we need to expose it?

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/DefaultObjectEntryManagerImpl.java [hunk 6]
```diff
@@ -1006,6 +1036,55 @@ private ServiceContext _createServiceContext(
...
 	}
 
+	private void _disassociateRelatedModel(
+			DTOConverterContext dtoConverterContext,
+			ObjectDefinition objectDefinition,
+			ObjectRelationship objectRelationship, long primaryKey,
+			ObjectDefinition relatedObjectDefinition)
+		throws Exception {
+
+		if (!relatedObjectDefinition.isUnmodifiableSystemObject()) {
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/DefaultObjectEntryManagerImpl.java:1046]: I think it is more clear in this case, if we have `if` and `else` to have a possitive clause

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/related/models/SystemObject1toMObjectRelatedModelsProviderImpl.java [hunk 7]
```diff
@@ -166,6 +169,70 @@ public void disassociateRelatedModels(
...
+		if (relatedObjectDefinition.isUnmodifiableSystemObject()) {
+			throw new UnsupportedOperationException();
+		}
+
+		PersistedModelLocalService persistedModelLocalService =
+			_persistedModelLocalServiceRegistry.getPersistedModelLocalService(
+				_systemObjectDefinitionManager.getModelClassName());
+
+		DynamicObjectDefinitionTable extensionDynamicObjectDefinitionTable =
+			_getExtensionDynamicObjectDefinitionTable(
```
Comment (ccorreagg) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/related/models/SystemObject1toMObjectRelatedModelsProviderImpl.java:194]: As we checked, we need to do the same as before. I mean, checking in the dynamic table and if not, in the extension dynamic table

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/ObjectEntry1toMObjectRelationshipElementsParserImpl.java (hunk 4 above)
Comment (ccorreagg) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/ObjectEntry1toMObjectRelationshipElementsParserImpl.java:57]: Maybe we could move it to parseOne. Something like:

```
@Override
protected ObjectEntry parseOne(Object object) {
	validateOne(object);

	Map<String, Object> objectMap = (Map<String, Object>)object;

	if (MapUtil.isEmpty(objectMap)) {
		return null;
	}

	return toObjectEntry((Map<String, Object>)object);
}
```

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/DefaultObjectEntryManagerImpl.java [hunk 8]
```diff
@@ -1000,6 +995,34 @@ private ServiceContext _createServiceContext(
...
+							objectRelationship.getType());
+
+				objectRelatedModelsProvider.disassociateRelatedModels(
+					dtoConverterContext.getUserId(),
+					objectRelationship.getObjectRelationshipId(),
+					objectEntry.getId(), primaryKey);
+			}
+		}
+	}
+
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/DefaultObjectEntryManagerImpl.java:1087]: In the method `dissasociateRelatedModels` in master I see it seems we were supporting many to one relationship. I think we should keep managing in in the same method

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/extension/v1_0/ObjectRelationshipExtensionProvider.java [hunk 9]
```diff
@@ -232,13 +233,39 @@ public void setExtendedProperties(
...
+
+				if (objectEntry != null) {
+					ObjectRelatedModelsProvider objectRelatedModelsProvider =
+						_objectRelatedModelsProviderRegistry.
+							getObjectRelatedModelsProvider(
+								objectDefinition.getClassName(),
+								objectDefinition.getCompanyId(),
+								objectRelationship.getType());
+
+					objectRelatedModelsProvider.disassociateRelatedModels(
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/extension/v1_0/ObjectRelationshipExtensionProvider.java:260]: Why don't we use the `defaultObjectEntryManager.disassociateRelatedModels` method instead?

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/DefaultObjectEntryManagerImpl.java [hunk 10]
```diff
@@ -237,27 +237,22 @@ public void deleteObjectEntry(
...
 	public void disassociateRelatedModels(
+			DTOConverterContext dtoConverterContext,
 			ObjectDefinition objectDefinition,
 			ObjectRelationship objectRelationship, long primaryKey,
 			ObjectDefinition relatedObjectDefinition, long userId)
 		throws Exception {
 
-		ObjectRelatedModelsProvider objectRelatedModelsProvider = null;
-
 		if (_isManyToOneObjectRelationship(
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/DefaultObjectEntryManagerImpl.java:246]: This is confusing. I don't see it clearly why we don't use this method

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/ObjectEntry1toMObjectRelationshipElementsParserImpl.java [hunk 11]
```diff
@@ -59,6 +66,12 @@ public List<ObjectEntry> parse(
 	protected ObjectEntry parseOne(Object object) {
 		validateOne(object);
 
+		Map<String, Object> objectMap = (Map<String, Object>)object;
+
+		if (MapUtil.isEmpty(objectMap)) {
+			return null;
+		}
+
 		return toObjectEntry((Map<String, Object>)object);
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/ObjectEntry1toMObjectRelationshipElementsParserImpl.java:75]: I guess we could use the already casted `objectMap`

**Review by ccorreagg — CHANGES_REQUESTED** [review]
Some comments and questions

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/extension/v1_0/ObjectRelationshipExtensionProvider.java [hunk 12]
```diff
@@ -232,16 +232,16 @@ public void setExtendedProperties(
...
-					relatedObjectDefinition, userId);
-			}
+			DefaultObjectEntryManager defaultObjectEntryManager =
+				DefaultObjectEntryManagerProvider.provide(
+					_objectEntryManagerRegistry.getObjectEntryManager(
+						objectDefinition.getStorageType()));
+
+			defaultObjectEntryManager.disassociateRelatedModels(
+				_getDefaultDTOConverterContext(
+					objectDefinition, getPrimaryKey(entity), null),
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/extension/v1_0/ObjectRelationshipExtensionProvider.java:242]: You are invoking many times to thje `getPrimaryKey` method. Please, move it to a variable

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/DefaultObjectEntryManagerImpl.java [hunk 13]
```diff
@@ -1101,6 +1127,34 @@ private Map<String, ObjectRelationship> _getObjectRelationships(
 		return objectRelationships;
 	}
 
+	private long _getRelatedModelPrimaryKey(
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/manager/v1_0/DefaultObjectEntryManagerImpl.java:1130]: The purpose of this method was to parse the Object passed by argument and return the primaryKey, not calling the fetch methods. You are duplicating code with the casting code (see `.getPrimaryKeyObj();`) and this method is called only once

**Comment by ccorreagg** [comment]
As we talked, I have pushed a couple of commits to simplify. forwarding...

**Comment by ccorreagg** [comment]
ci:forward

**Review by ccorreagg — APPROVED** [review]
LGTM!!!

**Comment by ccorreagg** [comment]
ci:forward

**Comment by sergiojimcos** [comment]
ci:forward

**Comment by ccorreagg** [comment]
ci:forward:force

**Pull request closed without merging**